File: 77 - Combinations/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    const result = [];

    const helper = (num, temp) => {

        if(temp.length == k){
            result.push([...temp]);
            return;
        }
        
        let i = num;
        while(i<=n){
            temp.push(i);
            helper(i+1, temp);
            temp.pop();
            i++;
        }
    }

    helper(1, []);
    return result;
};

File: 77 - Combinations/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public void helper(int num, int n, int k, List<Integer> temp, List<List<Integer>> result){

        if(temp.size() == k){
            result.add(new ArrayList<>(temp));
            return;
        }
        
        int i = num;
        while(i<=n){
            temp.add(i);
            helper(++i, n, k, temp, result);
            temp.remove(temp.size()-1);
        }
    }

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();

        helper(1, n, k, new ArrayList<>(), result);

        return result;    
    }
}